File: tests/unit/contrast.test.ts
/**
 * WCAG 1.4.3 (AA) contrast floor for the LIGHT palettes, read straight out of the stylesheet.
 *
 * Why this exists as a unit test rather than only in the browser: the light palettes shipped
 * with every brand hue failing AA — ember 2.7:1, amber 2.2:1, ice 3.1:1, on every page type —
 * and nothing caught it for six phases. The reason is mundane and worth writing down:
 * chrome-launcher inherits the host OS colour scheme, so every Lighthouse run this project
 * ever did on a dark-mode machine rendered the DARK palette and never once looked at the
 * light one. A gate that only runs a browser can be fooled by the machine it runs on; the
 * token values in the CSS cannot.
 *
 * The e2e suite (`tests/e2e/a11y.spec.ts`) does the complementary half — it pins
 * `data-theme` explicitly and measures what the pages actually render, catching pairs no one
 * thought to list here.
 */
import fs from 'node:fs';
import path from 'node:path';
import { describe, expect, it } from 'vitest';

const ROOT = path.resolve(import.meta.dirname, '..', '..');
const GLOBAL_CSS = fs.readFileSync(path.join(ROOT, 'src', 'styles', 'global.css'), 'utf8');
const REPO_CSS = fs.readFileSync(path.join(ROOT, 'src', 'styles', 'repo.css'), 'utf8');

/* ---- colour maths (the WCAG formulae, nothing more) ----------------------- */

type RGB = readonly [number, number, number];

function parseHex(value: string): RGB {
  const h = value.trim().replace('#', '');
  const full = h.length === 3 ? h.split('').map((c) => c + c).join('') : h;
  return [0, 2, 4].map((i) => Number.parseInt(full.slice(i, i + 2), 16)) as unknown as RGB;
}

/** `rgba(r, g, b, a)` → its channels plus alpha. */
function parseRgba(value: string): { rgb: RGB; alpha: number } {
  const m = /rgba?\(\s*([\d.]+)[,\s]+([\d.]+)[,\s]+([\d.]+)(?:[,\s/]+([\d.]+))?\s*\)/.exec(value);
  if (!m) throw new Error(`not an rgb(a) value: ${value}`);
  return {
    rgb: [Number(m[1]), Number(m[2]), Number(m[3])] as unknown as RGB,
    alpha: m[4] === undefined ? 1 : Number(m[4]),
  };
}

/** Flatten a translucent colour onto an opaque one — what the browser paints. */
function composite(fg: RGB, alpha: number, bg: RGB): RGB {
  return fg.map((v, i) => alpha * v + (1 - alpha) * bg[i]!) as unknown as RGB;
}

function relativeLuminance(c: RGB): number {
  const chan = (v: number) => {
    const s = v / 255;
    return s <= 0.04045 ? s / 12.92 : ((s + 0.055) / 1.055) ** 2.4;
  };
  return 0.2126 * chan(c[0]) + 0.7152 * chan(c[1]) + 0.0722 * chan(c[2]);
}

export function contrast(a: RGB, b: RGB): number {
  const [l1, l2] = [relativeLuminance(a), relativeLuminance(b)];
  return (Math.max(l1, l2) + 0.05) / (Math.min(l1, l2) + 0.05);
}

/* ---- reading the tokens out of the stylesheet ----------------------------- */

/** Every `--hf-*` declaration inside the block that starts with `selector {`. */
function tokenBlock(css: string, selector: string): Map<string, string> {
  const start = css.indexOf(`${selector} {`);
  if (start === -1) throw new Error(`no such block: ${selector}`);
  const open = css.indexOf('{', start);
  const end = css.indexOf('\n}', open);
  const body = css.slice(open + 1, end);
  const tokens = new Map<string, string>();
  for (const m of body.matchAll(/(--hf-[\w-]+)\s*:\s*([^;]+);/g)) tokens.set(m[1]!, m[2]!.trim());
  return tokens;
}

/** A palette's light tokens: the frost block only overrides, so it layers on the base. */
function lightPalette(name: 'hearth' | 'frost'): Map<string, string> {
  const base = tokenBlock(GLOBAL_CSS, ':root');
  if (name === 'hearth') return base;
  for (const [k, v] of tokenBlock(GLOBAL_CSS, ':root[data-palette="frost"]')) base.set(k, v);
  return base;
}

/* ---- the pairs that have to hold ------------------------------------------
 * Each entry is "this ink, on these backgrounds". The background list is not decorative:
 * it is where that colour is actually painted, checked against the stylesheet.
 */

/** WCAG AA for text below 18.66px bold / 24px regular — which every one of these is. */
const AA = 4.5;

interface Ink {
  /** Token name, or a literal for the two hard-coded diff colours. */
  ink: string;
  /** Token names of opaque backgrounds it is painted on. */
  on: string[];
  /**
   * Translucent tint tokens it is painted on, each flattened over `--hf-surface` and
   * `--hf-canvas` — a chip's tint sits on a card or on the page, and both have to clear.
   */
  onTints?: string[];
  why: string;
}

const INKS: Ink[] = [
  {
    ink: '--hf-ice',
    on: ['--hf-surface', '--hf-surface-2', '--hf-canvas', '--hf-canvas-2'],
    onTints: ['--hf-ice-soft'],
    why: '.hf-sha, .hf-age.t-cold, .hf-path-crumb links, .hf-tag, .hf-more, .hf-sha-link code',
  },
  {
    ink: '--hf-ember',
    on: ['--hf-surface', '--hf-surface-2', '--hf-canvas', '--hf-canvas-2'],
    onTints: ['--hf-ember-soft'],
    why: '.hf-age.t-hot, .hf-tag.hf-release-latest, .hf-tagkind--annotated',
  },
  {
    ink: '--hf-amber',
    on: ['--hf-surface', '--hf-surface-2', '--hf-canvas', '--hf-canvas-2'],
    onTints: ['--hf-amber-soft'],
    why: '.hf-footer-warn, .hf-release-pre, .hf-age.t-warm, .hf-tag--template',
  },
  {
    ink: '--hf-text-2',
    on: ['--hf-surface', '--hf-surface-2', '--hf-canvas', '--hf-canvas-2'],
    why: 'body copy in .hf-md, .hf-commit-author, .hf-blob-meta',
  },
  {
    ink: '--hf-text-3',
    on: ['--hf-surface', '--hf-surface-2', '--hf-canvas', '--hf-canvas-2'],
    onTints: ['--hf-code-bg'],
    why: '.hf-muted, table headers, .hf-kpi-sub, inline <code> in prose',
  },
];

describe.each(['hearth', 'frost'] as const)('light palette: %s', (name) => {
  const tokens = lightPalette(name);
  const rgb = (token: string) => parseHex(tokens.get(token)!);

  for (const entry of INKS) {
    it(`${entry.ink} clears AA where it is used (${entry.why})`, () => {
      const fg = rgb(entry.ink);
      for (const bg of entry.on) {
        expect(`${entry.ink} on ${bg}: ${contrast(fg, rgb(bg)).toFixed(2)}`).toBe(
          `${entry.ink} on ${bg}: ${Math.max(contrast(fg, rgb(bg)), AA).toFixed(2)}`,
        );
      }
      for (const tint of entry.onTints ?? []) {
        const { rgb: tintRgb, alpha } = parseRgba(tokens.get(tint)!);
        for (const under of ['--hf-surface', '--hf-canvas']) {
          const flattened = composite(tintRgb, alpha, rgb(under));
          const got = contrast(fg, flattened);
          expect(`${entry.ink} on ${tint} over ${under}: ${got.toFixed(2)}`).toBe(
            `${entry.ink} on ${tint} over ${under}: ${Math.max(got, AA).toFixed(2)}`,
          );
        }
      }
    });
  }

  /**
   * The diff colours are literals in repo.css rather than tokens — they are github's, not the
   * palette's — but they are still 12px text on a card and still have to clear AA on both
   * palettes. They live in `.hf-commit-row` / `.hf-commit-files`, whose backgrounds are the
   * card surface, its hover shade, and (on the single-commit head) the canvas.
   */
  it('diff +/- colours clear AA on both palettes', () => {
    const adds = /\.hf-adds \{ color: (#[0-9a-f]{6});/.exec(REPO_CSS)?.[1];
    const dels = /\.hf-dels \{ color: (#[0-9a-f]{6});/.exec(REPO_CSS)?.[1];
    expect(adds, '.hf-adds colour not found in repo.css').toBeDefined();
    expect(dels, '.hf-dels colour not found in repo.css').toBeDefined();
    for (const literal of [adds!, dels!]) {
      for (const bg of ['--hf-surface', '--hf-surface-2', '--hf-canvas']) {
        const got = contrast(parseHex(literal), rgb(bg));
        expect(`${literal} on ${bg}: ${got.toFixed(2)}`).toBe(
          `${literal} on ${bg}: ${Math.max(got, AA).toFixed(2)}`,
        );
      }
    }
  });

  /**
   * The code gutter is a `::before`, which means no browser-based checker will ever look at
   * it — axe and Lighthouse both skip pseudo-element text. It is still text, and it is still
   * `--hf-text-3` behind an `opacity`, so it is checked here or nowhere.
   */
  it('code line numbers clear AA through their opacity', () => {
    const opacity = Number(
      /\.hf-code \.line::before \{[\s\S]*?opacity: ([\d.]+);/.exec(REPO_CSS)?.[1] ?? 'NaN',
    );
    expect(opacity, 'line-number opacity not found in repo.css').toBeGreaterThan(0);
    for (const bg of ['--hf-surface', '--hf-surface-2']) {
      const painted = composite(rgb('--hf-text-3'), opacity, rgb(bg));
      const got = contrast(painted, rgb(bg));
      expect(`gutter on ${bg}: ${got.toFixed(2)}`).toBe(`gutter on ${bg}: ${Math.max(got, AA).toFixed(2)}`);
    }
  });

  /**
   * `--hf-accent` is the colour a primary button is painted IN, with `--hf-on-accent` written
   * on top of it. Darkening the accent for contrast in one direction must not break it in the
   * other.
   */
  it('button text clears AA on the accent fill', () => {
    const accent = tokens.get('--hf-accent')!.replace(/var\((--hf-[\w-]+)\)/, (_, t: string) => tokens.get(t)!);
    const onAccent = tokens.get('--hf-on-accent')!.replace(/var\((--hf-[\w-]+)\)/, (_, t: string) => tokens.get(t)!);
    const got = contrast(parseHex(accent), parseHex(onAccent));
    expect(`on-accent on accent: ${got.toFixed(2)}`).toBe(`on-accent on accent: ${Math.max(got, AA).toFixed(2)}`);
  });
});

/**
 * The DARK palettes were tuned when they were written and pass today; this pins them so a
 * future "let's unify the tokens" change cannot quietly undo that.
 */
describe.each([
  ['hearth', ':root[data-theme="dark"]'],
  ['frost', ':root[data-palette="frost"][data-theme="dark"]'],
] as const)('dark palette: %s', (_name, selector) => {
  const tokens = tokenBlock(GLOBAL_CSS, ':root');
  for (const [k, v] of tokenBlock(GLOBAL_CSS, selector)) tokens.set(k, v);
  const rgb = (token: string) => parseHex(tokens.get(token)!);

  for (const entry of INKS) {
    it(`${entry.ink} clears AA on the dark surfaces`, () => {
      const fg = rgb(entry.ink);
      for (const bg of entry.on) {
        const got = contrast(fg, rgb(bg));
        expect(`${entry.ink} on ${bg}: ${got.toFixed(2)}`).toBe(
          `${entry.ink} on ${bg}: ${Math.max(got, AA).toFixed(2)}`,
        );
      }
      for (const tint of entry.onTints ?? []) {
        const { rgb: tintRgb, alpha } = parseRgba(tokens.get(tint)!);
        for (const under of ['--hf-surface', '--hf-canvas']) {
          const got = contrast(fg, composite(tintRgb, alpha, rgb(under)));
          expect(`${entry.ink} on ${tint} over ${under}: ${got.toFixed(2)}`).toBe(
            `${entry.ink} on ${tint} over ${under}: ${Math.max(got, AA).toFixed(2)}`,
          );
        }
      }
    });
  }

  it('code line numbers clear AA through their opacity', () => {
    const opacity = Number(
      /\.hf-code \.line::before \{[\s\S]*?opacity: ([\d.]+);/.exec(REPO_CSS)?.[1] ?? 'NaN',
    );
    for (const bg of ['--hf-surface', '--hf-surface-2']) {
      const painted = composite(rgb('--hf-text-3'), opacity, rgb(bg));
      const got = contrast(painted, rgb(bg));
      expect(`gutter on ${bg}: ${got.toFixed(2)}`).toBe(`gutter on ${bg}: ${Math.max(got, AA).toFixed(2)}`);
    }
  });
});

File: internal/theme/contrast_test.go
package theme

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"testing"
)

// The WCAG AA floor for the palettes, read straight out of the stylesheets — the port of
// tests/unit/contrast.test.ts.
//
// Each entry below is "this ink, on these backgrounds". The background list is not decorative:
// it is where that colour is actually painted, and `why` names the rules that paint it, so a
// reader can check the claim rather than trust it.

type ink struct {
	// token is the CSS custom property, or a literal for the two hard-coded diff colours.
	token string
	// on are opaque background tokens it is painted on.
	on []string
	// onTints are translucent tint tokens it is painted on, each flattened over BOTH the card
	// surface and the page canvas — a chip's tint sits on either, and both have to clear.
	onTints []string
	why     string
}

var inks = []ink{
	{"--hf-ice", []string{"--hf-surface", "--hf-surface-2", "--hf-canvas", "--hf-canvas-2"}, []string{"--hf-ice-soft"},
		".hf-sha, .hf-age.t-cold, .hf-path-crumb links, .hf-tag, .hf-more, .hf-sha-link code"},
	{"--hf-ember", []string{"--hf-surface", "--hf-surface-2", "--hf-canvas", "--hf-canvas-2"}, []string{"--hf-ember-soft"},
		".hf-age.t-hot, .hf-tag.hf-release-latest, .hf-tagkind--annotated"},
	{"--hf-amber", []string{"--hf-surface", "--hf-surface-2", "--hf-canvas", "--hf-canvas-2"}, []string{"--hf-amber-soft"},
		".hf-footer-warn, .hf-release-pre, .hf-age.t-warm, .hf-tag--template"},
	{"--hf-text-2", []string{"--hf-surface", "--hf-surface-2", "--hf-canvas", "--hf-canvas-2"}, nil,
		"body copy in .hf-md, .hf-commit-author, .hf-blob-meta"},
	{"--hf-text-3", []string{"--hf-surface", "--hf-surface-2", "--hf-canvas", "--hf-canvas-2"}, []string{"--hf-code-bg"},
		".hf-muted, table headers, .hf-kpi-sub, inline <code> in prose"},
}

// palettes are the four token sets the site can render in: two palettes × two themes.
var palettes = []struct {
	name     string
	selector string // layered on top of :root; empty means :root alone
	dark     bool   // selects the dark overrides of the two literal diff colours
}{
	{"hearth light", "", false},
	{"frost light", `:root[data-palette="frost"]`, false},
	{"hearth dark", `:root[data-theme="dark"]`, true},
	{"frost dark", `:root[data-palette="frost"][data-theme="dark"]`, true},
}

func TestPalettesClearAA(t *testing.T) {
	globalCSS, repoCSS := stylesheets(t)

	for _, p := range palettes {
		t.Run(p.name, func(t *testing.T) {
			tokens := paletteTokens(t, globalCSS, p.selector)
			rgb := func(token string) RGB {
				t.Helper()
				v, ok := tokens[token]
				if !ok {
					t.Fatalf("no such token: %s", token)
				}
				c, err := ParseHex(Resolve(tokens, v))
				if err != nil {
					t.Fatalf("%s: %v", token, err)
				}
				return c
			}
			check := func(label string, got float64) {
				t.Helper()
				if got < AA {
					t.Errorf("%s: %.2f, below the AA floor of %.1f", label, got, AA)
				}
			}

			for _, in := range inks {
				fg := rgb(in.token)
				for _, bg := range in.on {
					check(fmt.Sprintf("%s on %s (%s)", in.token, bg, in.why), Contrast(fg, rgb(bg)))
				}
				for _, tint := range in.onTints {
					tintRGB, alpha, err := ParseRGBA(tokens[tint])
					if err != nil {
						t.Fatalf("%s: %v", tint, err)
					}
					for _, under := range []string{"--hf-surface", "--hf-canvas"} {
						flat := Composite(tintRGB, alpha, rgb(under))
						check(fmt.Sprintf("%s on %s over %s", in.token, tint, under), Contrast(fg, flat))
					}
				}
			}

			// The code gutter is a `::before`, so no browser-based checker will ever look at it —
			// axe and Lighthouse both skip pseudo-element text. It is still text, and still
			// --hf-text-3 behind an opacity, so it is checked here or nowhere.
			opacity := gutterOpacity(t, repoCSS)
			for _, bg := range []string{"--hf-surface", "--hf-surface-2"} {
				painted := Composite(rgb("--hf-text-3"), opacity, rgb(bg))
				check(fmt.Sprintf("code line numbers on %s", bg), Contrast(painted, rgb(bg)))
			}

			// The diff colours are literals in repo.css rather than tokens — they are github's,
			// not the palette's — but they are still 12px text on a card.
			for _, literal := range diffColours(t, repoCSS, p.dark) {
				c, err := ParseHex(literal)
				if err != nil {
					t.Fatal(err)
				}
				for _, bg := range []string{"--hf-surface", "--hf-surface-2", "--hf-canvas"} {
					check(fmt.Sprintf("diff colour %s on %s", literal, bg), Contrast(c, rgb(bg)))
				}
			}

			// --hf-accent is the colour a primary button is painted IN, with --hf-on-accent
			// written on top. Darkening the accent for contrast in one direction must not break
			// it in the other.
			check("--hf-on-accent on --hf-accent", Contrast(rgb("--hf-accent"), rgb("--hf-on-accent")))
		})
	}
}

// TestContrastMathsMatchesKnownValues pins the formulae themselves against values anyone can
// verify by hand, so a failure above is read as "the palette moved" and not "the maths is wrong".
func TestContrastMathsMatchesKnownValues(t *testing.T) {
	white, black := RGB{255, 255, 255}, RGB{0, 0, 0}
	if got := Contrast(white, black); got < 20.99 || got > 21.01 {
		t.Errorf("black on white = %.2f, want 21", got)
	}
	if got := Contrast(white, white); got < 0.99 || got > 1.01 {
		t.Errorf("white on white = %.2f, want 1", got)
	}
	// Half-transparent black over white is mid grey.
	if got := Composite(black, 0.5, white); got[0] < 127 || got[0] > 128 {
		t.Errorf("compositing gave %v, want ~127.5 per channel", got)
	}
	if c, err := ParseHex("#abc"); err != nil || c != (RGB{0xAA, 0xBB, 0xCC}) {
		t.Errorf("three-digit hex: %v %v", c, err)
	}
}

func stylesheets(t *testing.T) (globalCSS, repoCSS string) {
	t.Helper()
	root := moduleRoot(t)
	read := func(name string) string {
		raw, err := os.ReadFile(filepath.Join(root, "web", "css", name))
		if err != nil {
			t.Fatalf("read %s: %v", name, err)
		}
		return string(raw)
	}
	return read("global.css"), read("repo.css")
}

func paletteTokens(t *testing.T, css, selector string) map[string]string {
	t.Helper()
	base, err := TokenBlock(css, ":root")
	if err != nil {
		t.Fatal(err)
	}
	if selector == "" {
		return base
	}
	over, err := TokenBlock(css, selector)
	if err != nil {
		t.Fatal(err)
	}
	return Layer(base, over)
}

var (
	gutterRe = regexp.MustCompile(`(?s)\.hf-code \.line::before \{.*?opacity: ([\d.]+);`)
	addsRe   = regexp.MustCompile(`(?m)^\.hf-adds \{ color: (#[0-9a-f]{6});`)
	delsRe   = regexp.MustCompile(`(?m)^\.hf-dels \{ color: (#[0-9a-f]{6});`)
	// The dark theme repaints both, so checking the light literals against dark surfaces would
	// invent a failure that no reader ever sees.
	darkAddsRe = regexp.MustCompile(`:root\[data-theme="dark"\] \.hf-adds \{ color: (#[0-9a-f]{6});`)
	darkDelsRe = regexp.MustCompile(`:root\[data-theme="dark"\] \.hf-dels \{ color: (#[0-9a-f]{6});`)
)

func gutterOpacity(t *testing.T, repoCSS string) float64 {
	t.Helper()
	m := gutterRe.FindStringSubmatch(repoCSS)
	if m == nil {
		t.Fatal("line-number opacity not found in repo.css — the gutter rule moved, and this check went quiet with it")
	}
	v, err := strconv.ParseFloat(m[1], 64)
	if err != nil || v <= 0 {
		t.Fatalf("line-number opacity %q is not a positive number", m[1])
	}
	return v
}

func diffColours(t *testing.T, repoCSS string, dark bool) []string {
	t.Helper()
	a, d := addsRe, delsRe
	if dark {
		a, d = darkAddsRe, darkDelsRe
	}
	adds := a.FindStringSubmatch(repoCSS)
	dels := d.FindStringSubmatch(repoCSS)
	if adds == nil || dels == nil {
		t.Fatalf(".hf-adds / .hf-dels colours not found in repo.css (dark=%v) — the rules moved, and this check went quiet with them", dark)
	}
	return []string{adds[1], dels[1]}
}

func moduleRoot(t *testing.T) string {
	t.Helper()
	dir, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	for i := 0; i < 6; i++ {
		if _, err := os.Stat(filepath.Join(dir, "go.mod")); err == nil {
			return dir
		}
		dir = filepath.Dir(dir)
	}
	t.Fatal("could not find the module root")
	return ""
}
